### PR TITLE
JSDOM port

### PR DESCRIPTION
**Pull request by MarkTiedemann**
Closes https://github.com/denoland/deno_std/issues/537.

```
$ cd jsdom
$ npm install --no-package-lock webpack@4.36.1 webpack-cli@3.3.6 jsdom@15.1.1
$ ./node_modules/.bin/webpack-cli
Hash: 80cb94c06fc1cfb9a48b
Version: webpack 4.36.1
Time: 2422ms
Built at: 07/20/2019 11:51:39 PM
     Asset      Size  Chunks                    Chunk Names
./jsdom.js  3.05 MiB       0  [emitted]  [big]  main
Entrypoint main [big] = ./jsdom.js
 [17] (webpack)/buildin/global.js 472 bytes {0} [built]
[108] (webpack)/buildin/module.js 497 bytes {0} [built]
[231] canvas (ignored) 15 bytes {0} [built]
[440] ./streams (ignored) 15 bytes {0} [built]
[441] ./extend-node (ignored) 15 bytes {0} [built]
[519] util (ignored) 15 bytes {0} [built]
[521] util (ignored) 15 bytes {0} [built]
[562] buffer (ignored) 15 bytes {0} [optional] [built]
[563] crypto (ignored) 15 bytes {0} [optional] [built]
[625] crypto (ignored) 15 bytes {0} [built]
    + 948 hidden modules
$ deno jsdom.ts
Compile file:///Users/marktiedemann/dev/deno_std/jsdom/jsdom.ts
Hello, Deno!
```

TODOs:

- [x] Add preliminary types.
- [x] Add preliminary tests.
- [x] Write documentation.

Looking for feedback.

cc @ry

### CODE REVIEW

File: jsdom/jsdom.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import "./vendor/jsdom.js";
+
+let JSDOM = (window as any).jsdom.JSDOM;
+let dom = new JSDOM(`
+  <!DOCTYPE html>
+  <p>Hello, Deno!</p>
+`);
```
Comment (ry) [inline jsdom/jsdom.ts]: Cool! How big is the vendor/jsdom.js file?

Comment (MarkTiedemann) [reply jsdom/jsdom.ts]: 3.05 MiB

I added a TODO for investigating whether we can reduce this size.

File: jsdom/package.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "scripts": {
+    "vendor": "webpack-cli"
+  },
+  "devDependencies": {
+    "jsdom": "15.1.1",
+    "webpack": "4.36.1",
+    "webpack-cli": "3.3.6"
+  }
+}
```
Comment (MarkTiedemann) [inline jsdom/package.json]: I added the `package.json` so you can reproduce building the vendor bundle.

Is this a good approach, @ry?

Might be something we can do for `//deno_std/prettier`, too.

File: jsdom/jsdom.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import "./vendor/jsdom.js";
+
+// TODO: Investigate how to reuse typings from:
+// - https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jsdom/index.d.ts
+// - https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts
```
Comment (MarkTiedemann) [inline jsdom/jsdom.ts]: @kitsonk Any idea on how to reuse these types here?

Comment (kitsonk) [reply jsdom/jsdom.ts]: There isn't really a way at the moment.  It is covered by denoland/deno#1432.

Comment (MarkTiedemann) [reply jsdom/jsdom.ts]: Ah, too bad. I guess I will add some preliminary types for now, and we can improve them later, once `.d.ts` files are supported.

**Comment by MarkTiedemann** [comment]
I think this is ready to be reviewed now, @ry.

One major painpoint, I guess, is that the DOM and JSDOM types are still preliminary. There are also some other future TODOs, for example, investigating whether the bundle size can be improved, testing more of JSDOM's functionality, and rewriting/wrapping some Node specific-parts in JSDOM (for example, `JSDOM.fromURL` and `JSDOM.fromFile` both use built-in Node modules). But I think these can be tackled in future PRs.

File: jsdom/mod.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+import { jsdom as instance } from "./vendor/jsdom.js";
+import { jsdom } from "./types/jsdom.ts";
+
+export const JSDOM = instance.JSDOM as jsdom.JSDOM;
```
Comment (j-f1) [inline jsdom/mod.ts]: Another implementation of this that should also work:

```js
export { jsdom as JSDOM } from "./vendor/jsdom.js";
export { jsdom as JSDOM } from "./types/jsdom.ts";

Comment (MarkTiedemann) [reply jsdom/mod.ts]: Thanks for the suggestion, @j-f1. Unfortunately, this doesn't work on my machine. When I change it, `JSDOM` can't be imported anymore.

```
$ deno test.ts
Compile file:///Users/marktiedemann/dev/deno_std/jsdom/mod.ts
error: Uncaught SyntaxError: The requested module './mod.ts' does not provide an export named 'JSDOM'
```

Comment (j-f1) [reply jsdom/mod.ts]: Well, that’s unexpected. Sorry to send you on a wild goose chase!

Comment (MarkTiedemann) [reply jsdom/mod.ts]: No worries, @j-f1.

I just tried again, with the same code, and got a different error:

```
$ deno test.ts
Compile file:///Users/marktiedemann/dev/deno_std/jsdom/mod.ts

error TS2300: Duplicate identifier 'JSDOM'.

► file:///Users/marktiedemann/dev/deno_std/jsdom/mod.ts:1:19

1 export { jsdom as JSDOM } from "./vendor/jsdom.js";
                    ~~~~~

error TS2300: Duplicate identifier 'JSDOM'.

► file:///Users/marktiedemann/dev/deno_std/jsdom/mod.ts:2:19

2 export { jsdom as JSDOM } from "./types/jsdom.ts";
                    ~~~~~


Found 2 errors.
```

This seems like a more expected error, still unfortunate.

Comment (kitsonk) [reply jsdom/mod.ts]: I suspect denoland/deno#2746 will fix this (or at least make it possible).

Comment (MarkTiedemann) [reply jsdom/mod.ts]: @kitsonk, awesome! As soon as https://github.com/denoland/deno/pull/2746 is released, I will update the types here.

**Comment by ry** [comment]
I'm having trouble running the tests with the big bundle file...

**Pull request closed without merging**